File: RxBus/src/main/java/com/adiliqbal/rxbus/Optional.java
package com.adiliqbal.rxbus;


public class Optional<T> {

    private T value;

    private Optional() {
        this.value = null;
    }

    private Optional(T value) {
        this.value = value;
    }

    public static <T> Optional<T> empty() {
        return new Optional<>();
    }

    public static <T> Optional<T> of(T value) {
        return new Optional<>(value);
    }

    public T get() {
        return value;
    }

    public boolean isPresent() {
        return value != null;
    }
}

File: RxBus/src/main/java/com/adiliqbal/rxbus/Optional.kt
package com.adiliqbal.rxbus

class Optional<T> {

    private var value: T? = null

    val isPresent: Boolean
        get() = value != null

    private constructor() {
        this.value = null
    }

    private constructor(value: T) {
        this.value = value
    }

    fun get(): T? {
        return value
    }

    companion object {

        fun <T> empty(): Optional<T> {
            return Optional()
        }

        fun <T> of(value: T): Optional<T> {
            return Optional(value)
        }
    }
}
